pkg/validation/internal/object.go

```diff
@@ -0,0 +1,199 @@
+package internal
+
+import (
+	"encoding/json"
+	"github.com/operator-framework/api/pkg/validation/errors"
+	interfaces "github.com/operator-framework/api/pkg/validation/interfaces"
+
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	schedulingv1 "k8s.io/api/scheduling/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+var ObjectValidator interfaces.Validator = interfaces.ValidatorFunc(validateObjects)
+
+const (
+	PodDisruptionBudgetKind     = "PodDisruptionBudget"
+	PriorityClassKind           = "PriorityClass"
+	RoleKind                    = "Role"
+	ClusterRoleKind             = "ClusterRole"
+	PodDisruptionBudgetAPIGroup = "policy"
+	SCCAPIGroup                 = "security.openshift.io"
+)
+
+var defaultSCCs = map[string]struct{}{
+	"privileged":       {},
+	"restricted":       {},
+	"anyuid":           {},
+	"hostaccess":       {},
+	"hostmount-anyuid": {},
+	"hostnetwork":      {},
+	"node-exporter":    {},
+	"nonroot":          {},
+}
+
+func validateObjects(objs ...interface{}) (results []errors.ManifestResult) {
+	for _, obj := range objs {
+		switch u := obj.(type) {
+		case *unstructured.Unstructured:
+			switch u.GroupVersionKind().Kind {
+			case PodDisruptionBudgetKind:
+				results = append(results, validatePDB(u))
+			case PriorityClassKind:
+				results = append(results, validatePriorityClass(u))
+			case RoleKind:
+				results = append(results, validateRBAC(u))
+			case ClusterRoleKind:
+				results = append(results, validateRBAC(u))
+			}
+		}
+	}
+	return results
+}
+
+// validatePDB checks the PDB to ensure the minimum and maximum budgets are set to reasonable levels.
+// See https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/adding-pod-disruption-budgets.md#limitations-on-pod-disruption-budgets
+func validatePDB(u *unstructured.Unstructured) (result errors.ManifestResult) {
+	pdb := policyv1beta1.PodDisruptionBudget{}
+
+	b, err := u.MarshalJSON()
+	if err != nil {
+		result.Add(errors.ErrInvalidParse("error converting unstructured", err))
+		return
+	}
+
+	err = json.Unmarshal(b, &pdb)
+	if err != nil {
+		result.Add(errors.ErrInvalidParse("error unmarshaling poddisruptionbudget", err))
+		return
+	}
+
+	/*
+	   maxUnavailable field cannot be set to 0 or 0%.
+	   minAvailable field cannot be set to 100%.
+	*/
+
+	maxUnavailable := pdb.Spec.MaxUnavailable
+	if maxUnavailable != nil {
+		if maxUnavailable.IntVal == 0 || maxUnavailable.StrVal == "0%" {
+			result.Add(errors.ErrInvalidObject(pdb, "maxUnavailable field cannot be set to 0 or 0%"))
+		}
+	}
+
+	minAvailable := pdb.Spec.MinAvailable
+	if minAvailable != nil {
+		if minAvailable.StrVal == "100%" {
```

Comment:
```suggestion
		if minAvailable.IntVal == 100 || minAvailable.StrVal == "100%" {
```

Reply:
If the manifest says `MinAvailable: 100` then it means they want 100 pods always available --- this is strange but not equivalent to 100% of pods available right? I don't think we need to block on the `IntVal` here, unlike in the `maxUnavailable` case.

Reply:
If `MinAvailable: 100` is on the spec it means keep at least 100 pods on the cluster -- this is strange but necessarily the same as 100% of pods. I don't think we need to block on the absolute number here, unlike in the `maxUnavailable` case where 0 and 0% are equivalent.